File: experiments/java-cpp-matching/cpp/matching_lab.cpp
// 单线程、单交易对的内存撮合实验；不提供 HTTP、账本、持久化或生产替换入口。
// price 单位为 1/10000 报价币，quantity 为整数最小交易单位。禁止浮点金额。
// 标准容器版本刻意保留真实分配成本：树节点、链表节点、订单、请求与成交结果均会分配。
#include <algorithm>
#include <charconv>
#include <chrono>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <iterator>
#include <limits>
#include <list>
#include <map>
#include <memory>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
#include <unistd.h>

using Amount = std::int64_t;
static constexpr Amount MAX_AMOUNT = std::numeric_limits<Amount>::max();
static void require(bool condition, const char* message) { if (!condition) throw std::runtime_error(message); }
static bool identifier(const std::string& s) {
    if (s.empty() || s.size() > 40) return false;
    return std::all_of(s.begin(), s.end(), [](char c) { return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_'; });
}
static bool parseInteger(const std::string& s, Amount& value) {
    if (s.empty() || s[0] == '+') return false;
    const char* first = s.data(); const char* last = first + s.size();
    auto result = std::from_chars(first, last, value);
    return result.ec == std::errc{} && result.ptr == last;
}

struct Command {
    bool valid = false, badNumber = false;
    std::string fingerprint, kind, request = "?", id, owner, side;
    Amount price = 0, quantity = 0;
    // 解析和输入分配均在测量区间外，指纹规范化空白但不改变数值文本。
    static Command parse(const std::string& line) {
        std::istringstream in(line); std::vector<std::string> t;
        for (std::string word; in >> word;) t.push_back(std::move(word));
        Command c;
        if (!((t.size() == 7 && t[0] == "N") || (t.size() == 4 && t[0] == "C"))) return c;
        for (std::size_t i = 1; i <= 3; ++i) if (!identifier(t[i])) return c;
        c.valid = true; c.kind = t[0]; c.request = t[1]; c.id = t[2]; c.owner = t[3];
        for (const auto& field : t) { if (!c.fingerprint.empty()) c.fingerprint += ' '; c.fingerprint += field; }
        if (c.kind == "N") { c.side = t[4]; c.badNumber = !parseInteger(t[5], c.price) || !parseInteger(t[6], c.quantity); }
        return c;
    }
};

struct Order {
    std::string id, owner, side;
    Amount price, original, remaining, filled = 0, cancelled = 0, executed = 0;
    std::uint64_t sequence;
    std::list<Order*>::iterator position;
    Order(const Command& c, std::uint64_t seq) : id(c.id), owner(c.owner), side(c.side), price(c.price), original(c.quantity), remaining(c.quantity), sequence(seq) {}
};
struct Trade { std::string maker; Amount quantity, price; };
struct Result {
    std::string kind, request, id;
    Amount remaining = 0, cancelled = 0;
    std::vector<Trade> trades;
    std::string error;
    static Result reject(const std::string& request, const std::string& error) { return {"R", request, "", 0, 0, {}, error}; }
    std::string encode() const {
        if (kind == "R") return "R|" + request + "|" + error;
        if (kind == "C") return "C|" + request + "|" + id + "|" + std::to_string(cancelled);
        std::ostringstream out; out << "A|" << request << '|' << id << '|' << remaining << '|';
        for (std::size_t i = 0; i < trades.size(); ++i) { if (i) out << ','; const auto& t = trades[i]; out << t.maker << ':' << t.quantity << ':' << t.price; }
        if (trades.empty()) out << '-';
        out << '|' << cancelled; return out.str();
    }
};
struct Cached { std::string fingerprint; Result result; };

class Engine {
    // 与 Java 对应：有序价位 + FIFO 队列 + 订单哈希索引。撤单使用稳定链表迭代器。
    using Book = std::map<Amount, std::list<Order*>>;
    Book bids, asks;
    std::unordered_map<std::string, std::unique_ptr<Order>> orders;
    std::unordered_map<std::string, Cached> requests;
    std::uint64_t sequence = 0;
    Result transient;
    Book& book(const Order& o) { return o.side == "B" ? bids : asks; }
    void remove(Order& o) {
        Book& b = book(o); auto level = b.find(o.price);
        level->second.erase(o.position); if (level->second.empty()) b.erase(level);
    }
    Result cancel(const Command& c) {
        auto found = orders.find(c.id);
        if (found == orders.end()) return Result::reject(c.request, "NOT_FOUND");
        Order& o = *found->second;
        if (o.owner != c.owner) return Result::reject(c.request, "NOT_OWNER");
        if (o.remaining == 0) return Result::reject(c.request, "TERMINAL");
        Amount cancelled = o.remaining;
        remove(o); o.cancelled += cancelled; o.remaining = 0;
        return {"C", c.request, c.id, 0, cancelled, {}, ""};
    }
    Result place(const Command& c) {
        if (c.side != "B" && c.side != "S") return Result::reject(c.request, "BAD_SIDE");
        if (c.badNumber) return Result::reject(c.request, "BAD_NUMBER");
        if (c.price <= 0 || c.quantity <= 0) return Result::reject(c.request, "NON_POSITIVE");
        // 不能先乘再检测：C++ 有符号整数溢出是未定义行为。先除法比较，再进行任何状态修改。
        if (c.price > MAX_AMOUNT / c.quantity) return Result::reject(c.request, "OVERFLOW");
        if (orders.contains(c.id)) return Result::reject(c.request, "ORDER_EXISTS");
        if (executionWouldOverflow(c)) return Result::reject(c.request, "OVERFLOW");
        auto holder = std::make_unique<Order>(c, sequence++); Order& taker = *holder;
        Book& opposite = c.side == "B" ? asks : bids;
        std::vector<Trade> trades;
        while (taker.remaining > 0 && !opposite.empty()) {
            auto level = c.side == "B" ? opposite.begin() : std::prev(opposite.end());
            Amount price = level->first;
            if (c.side == "B" ? price > c.price : price < c.price) break;
            Order& maker = *level->second.front();
            // CANCEL_TAKER：保留本命令此前有效成交，取消剩余量，自有 maker 保持不变。
            if (maker.owner == c.owner) { taker.cancelled = taker.remaining; taker.remaining = 0; break; }
            Amount quantity = std::min(taker.remaining, maker.remaining);
            maker.remaining -= quantity; maker.filled += quantity;
            taker.remaining -= quantity; taker.filled += quantity;
            Amount notional = quantity * maker.price;
            maker.executed += notional; taker.executed += notional;
            trades.push_back({maker.id, quantity, maker.price});
            if (maker.remaining == 0) remove(maker);
        }
        orders.emplace(taker.id, std::move(holder));
        if (taker.remaining > 0) { auto& queue = book(taker)[taker.price]; queue.push_back(&taker); taker.position = std::prev(queue.end()); }
        return {"A", c.request, c.id, taker.remaining, taker.cancelled, std::move(trades), ""};
    }
    // 卖出价改善可能让累计成交额超过限价*数量。预检实际路径并为残量预留限价成交额。
    // 这次只读遍历先于任何状态修改；价位/订单扫描两次的 CPU 成本被双方计时包含。
    bool executionWouldOverflow(const Command& c) const {
        Amount left = c.quantity, total = 0; bool overflow = false;
        auto visit = [&](const auto& level) {
            if (c.side == "B" ? level.first > c.price : level.first < c.price) return false;
            for (const Order* maker : level.second) {
                if (maker->owner == c.owner) { left = 0; return false; }
                Amount quantity = std::min(left, maker->remaining), notional = quantity * maker->price;
                if (total > MAX_AMOUNT - notional) { overflow = true; return false; }
                total += notional; left -= quantity; if (left == 0) return false;
            }
            return true;
        };
        if (c.side == "B") { for (const auto& level : asks) if (!visit(level)) break; }
        else { for (auto it = bids.rbegin(); it != bids.rend(); ++it) if (!visit(*it)) break; }
        return overflow || total > MAX_AMOUNT - left * c.price;
    }
    void checkBook(const Book& b, const std::string& side, std::unordered_set<const Order*>& seen) const {
        for (const auto& [price, queue] : b) {
            require(!queue.empty(), "empty level"); std::uint64_t previous = 0; bool first = true;
            for (auto iterator = queue.begin(); iterator != queue.end(); ++iterator) {
                const Order& o = **iterator;
                require(o.price == price && o.side == side && o.remaining > 0, "bad level");
                require((first || o.sequence > previous) && seen.insert(&o).second, "FIFO/duplicate index");
                require(orders.at(o.id).get() == &o && o.position == iterator, "order index/iterator");
                previous = o.sequence; first = false;
            }
        }
    }
public:
    const Result& apply(const Command& c) {
        if (!c.valid) { transient = Result::reject("?", "BAD_PROTOCOL"); return transient; }
        auto previous = requests.find(c.request);
        if (previous != requests.end()) {
            if (previous->second.fingerprint == c.fingerprint) return previous->second.result;
            transient = Result::reject(c.request, "DUPLICATE_KEY"); return transient;
        }
        Result result = c.kind == "C" ? cancel(c) : place(c);
        return requests.emplace(c.request, Cached{c.fingerprint, std::move(result)}).first->second.result;
    }
    void verify() const {
        std::unordered_set<const Order*> seen; checkBook(bids, "B", seen); checkBook(asks, "S", seen);
        for (const auto& [id, holder] : orders) {
            const Order& o = *holder;
            require(o.remaining >= 0 && o.filled >= 0 && o.cancelled >= 0, "negative quantity");
            require(o.original - o.remaining - o.filled == o.cancelled, "quantity conservation");
            require(o.executed >= 0 && o.executed <= MAX_AMOUNT - o.remaining * o.price, "execution notional overflow");
            require((o.remaining > 0) == seen.contains(&o), "order/book disagreement");
        }
        require(bids.empty() || asks.empty() || bids.rbegin()->first < asks.begin()->first, "crossed book");
    }
    std::string snapshot() const {
        std::vector<const Order*> sorted; for (const auto& [id, holder] : orders) sorted.push_back(holder.get());
        std::sort(sorted.begin(), sorted.end(), [](const Order* a, const Order* b) { return a->id < b->id; });
        std::ostringstream out; bool first = true;
        for (const Order* o : sorted) { if (!first) out << ';'; first = false; out << o->id << ':' << o->owner << ':' << o->side << ':' << o->price << ':' << o->original << ':' << o->remaining << ':' << o->filled << ':' << o->cancelled << ':' << o->executed; }
        return out.str();
    }
};

static std::string hash(const std::string& state) {
    // 状态散列使用有意定义的无符号模运算，不参与金额运算。
    std::uint64_t value = 14695981039346656037ULL;
    for (unsigned char c : state) value = (value ^ c) * 1099511628211ULL;
    return std::to_string(value);
}
static std::int64_t percentile(const std::vector<std::int64_t>& sorted, std::size_t numerator, std::size_t denominator) {
    return sorted.at((sorted.size() * numerator + denominator - 1) / denominator - 1);
}
static volatile std::size_t observable = 0;

static int execute(int argc, char** argv) {
    if (argc == 2 && (std::string(argv[1]) == "--replay" || std::string(argv[1]) == "--replay-batch")) {
        Engine engine;
        for (std::string line; std::getline(std::cin, line);) { std::cout << engine.apply(Command::parse(line)).encode() << '\n'; if (std::string(argv[1]) == "--replay") engine.verify(); }
        engine.verify(); std::cout << "STATE|" << engine.snapshot() << '\n'; return 0;
    }
    require(argc == 4 && std::string(argv[1]) == "--bench", "--replay | --bench FILE WARMUP");
    std::ifstream input(argv[2]); require(input.good(), "cannot open workload");
    std::vector<Command> commands;
    for (std::string line; std::getline(input, line);) commands.push_back(Command::parse(line));
    int warmup = std::stoi(argv[3]); require(!commands.empty() && warmup > 0, "empty workload or missing warmup");
    for (int pass = 0; pass < warmup; ++pass) { Engine warm; for (const auto& c : commands) warm.apply(c); warm.verify(); observable = observable ^ warm.snapshot().size(); }
    Engine engine; std::vector<std::int64_t> samples(commands.size());
    using Clock = std::chrono::steady_clock;
    auto started = Clock::now();
    for (std::size_t i = 0; i < commands.size(); ++i) {
        auto before = Clock::now(); engine.apply(commands[i]);
        samples[i] = std::chrono::duration_cast<std::chrono::nanoseconds>(Clock::now() - before).count();
    }
    auto elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(Clock::now() - started).count();
    engine.verify(); std::string stateHash = hash(engine.snapshot()); observable = observable ^ stateHash.size();
    std::string samplePath = std::string(argv[2]) + ".cpp.samples." + std::to_string(getpid()) + ".csv";
    { std::ofstream output(samplePath); require(output.good(), "cannot write samples"); output << "latency_ns\n"; for (auto sample : samples) output << sample << '\n'; require(output.good(), "sample write failure"); }
    std::sort(samples.begin(), samples.end());
    auto filename = samplePath.substr(samplePath.find_last_of("/\\") + 1);
    // 唯一浮点表达式用于吞吐率（时间统计），不用于任何价格、数量或金额。
    std::cout << "{\"samples\":" << samples.size() << ",\"elapsedNs\":" << elapsed << ",\"throughputOpsSec\":" << (samples.size() * 1e9 / static_cast<double>(elapsed))
              << ",\"p50Ns\":" << percentile(samples, 1, 2) << ",\"p99Ns\":" << percentile(samples, 99, 100) << ",\"p999Ns\":" << percentile(samples, 999, 1000)
              << ",\"maxNs\":" << samples.back() << ",\"allocatedBytesMeasuredThread\":null,\"stateHash\":\"" << stateHash << "\",\"sampleFile\":\"" << filename << "\"}\n";
    return 0;
}
int main(int argc, char** argv) {
    // 内存分配失败等异常终止进程；不能把部分执行伪装成成功，也没有持久化恢复保证。
    try { return execute(argc, argv); }
    catch (const std::exception& error) { std::cerr << "matching-lab failure: " << error.what() << '\n'; return 2; }
}

File: experiments/java-cpp-matching/java/MatchingLab.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.lang.management.ManagementFactory;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;
import java.util.regex.Pattern;

/**
 * 独立、单写线程、单交易对的内存撮合实验。它不是账本，也没有生产 API。
 * 金额契约：price 为 1/10000 报价币的整数 tick；quantity 为整数最小交易单位。
 * 所有历史订单和请求保留到本进程结束，便于证明终态不复活和重试不重复执行。
 * 生产迁移必须另外实现持久化、分区 fencing、恢复、清理策略和事务边界。
 *
 * @author FinCore Reliability Lab
 * @since 2026-09-06
 */
public final class MatchingLab {
    private static final Pattern ID = Pattern.compile("[A-Za-z0-9_]{1,40}");
    private static final Pattern INTEGER = Pattern.compile("-?[0-9]+");

    /** 文本解析在计时外；保留规范化文本作为请求指纹，字段不同即冲突。 */
    record Command(boolean valid, String fingerprint, String kind, String request,
                   String id, String owner, String side, long price, long quantity, boolean badNumber) {
        static Command parse(String line) {
            String[] t = line.strip().split("\\s+");
            boolean valid = (t.length == 7 && t[0].equals("N")) || (t.length == 4 && t[0].equals("C"));
            if (valid) {
                for (int i = 1; i <= 3; i++) {
                    valid &= ID.matcher(t[i]).matches();
                }
            }
            if (!valid) {
                return new Command(false, "", "", "?", "", "", "", 0, 0, false);
            }
            long price = 0, quantity = 0; boolean badNumber = false;
            if (t[0].equals("N")) {
                try {
                    if (!INTEGER.matcher(t[5]).matches() || !INTEGER.matcher(t[6]).matches()) {
                        throw new NumberFormatException();
                    }
                    price = Long.parseLong(t[5]); quantity = Long.parseLong(t[6]);
                } catch (NumberFormatException expectedInvalidInput) {
                    badNumber = true;
                }
            }
            return new Command(true, String.join(" ", t), t[0], t[1], t[2], t[3], t.length == 7 ? t[4] : "", price, quantity, badNumber);
        }
    }

    /** 对象字段记录守恒关系；FIFO 由价位内 LinkedHashMap 插入顺序决定。 */
    static final class Order {
        final String id, owner, side;
        final long price, original, sequence;
        long remaining, filled, cancelled, executed;
        Order(Command c, long sequence) {
            id = c.id; owner = c.owner; side = c.side; price = c.price;
            original = c.quantity; remaining = c.quantity; this.sequence = sequence;
        }
    }
    record Trade(String maker, long quantity, long price) {}
    record Result(String kind, String request, String id, long remaining, long cancelled, List<Trade> trades, String error) {
        static Result reject(String request, String code) { return new Result("R", request, "", 0, 0, List.of(), code); }
        String encode() {
            if (kind.equals("R")) {
                return "R|" + request + "|" + error;
            }
            if (kind.equals("C")) {
                return "C|" + request + "|" + id + "|" + cancelled;
            }
            StringBuilder b = new StringBuilder("A|").append(request).append('|').append(id).append('|').append(remaining).append('|');
            for (Trade t : trades) {
                if (b.charAt(b.length() - 1) != '|') {
                    b.append(',');
                }
                b.append(t.maker).append(':').append(t.quantity).append(':').append(t.price);
            }
            if (trades.isEmpty()) {
                b.append('-');
            }
            return b.append('|').append(cancelled).toString();
        }
    }
    record Cached(String fingerprint, Result result) {}

    /** 串行状态机；业务拒绝发生在修改前，进程失败没有恢复或事务承诺。 */
    static final class Engine {
        // 红黑树定位最优价格 O(log P)，价位内插入/撤单平均 O(1)。对象分配与指针追踪均在测量范围。
        final NavigableMap<Long, LinkedHashMap<String, Order>> bids = new TreeMap<>(Comparator.reverseOrder());
        final NavigableMap<Long, LinkedHashMap<String, Order>> asks = new TreeMap<>();
        final Map<String, Order> orders = new HashMap<>();
        final Map<String, Cached> requests = new HashMap<>();
        long sequence;

        Result apply(Command c) {
            if (!c.valid) {
                return Result.reject("?", "BAD_PROTOCOL");
            }
            Cached previous = requests.get(c.request);
            if (previous != null) {
                return previous.fingerprint.equals(c.fingerprint) ? previous.result : Result.reject(c.request, "DUPLICATE_KEY");
            }
            Result result = c.kind.equals("C") ? cancel(c) : place(c);
            requests.put(c.request, new Cached(c.fingerprint, result));
            return result;
        }

        Result cancel(Command c) {
            Order o = orders.get(c.id);
            if (o == null) {
                return Result.reject(c.request, "NOT_FOUND");
            }
            if (!o.owner.equals(c.owner)) {
                return Result.reject(c.request, "NOT_OWNER");
            }
            if (o.remaining == 0) {
                return Result.reject(c.request, "TERMINAL");
            }
            long cancelled = o.remaining;
            remove(o); o.cancelled += cancelled; o.remaining = 0;
            return new Result("C", c.request, c.id, 0, cancelled, List.of(), "");
        }

        Result place(Command c) {
            if (!c.side.equals("B") && !c.side.equals("S")) {
                return Result.reject(c.request, "BAD_SIDE");
            }
            if (c.badNumber) {
                return Result.reject(c.request, "BAD_NUMBER");
            }
            if (c.price <= 0 || c.quantity <= 0) {
                return Result.reject(c.request, "NON_POSITIVE");
            }
            // 除法式上界检查先于任何订单/价位修改，避免溢出后部分成交。
            if (c.price > Long.MAX_VALUE / c.quantity) {
                return Result.reject(c.request, "OVERFLOW");
            }
            if (orders.containsKey(c.id)) {
                return Result.reject(c.request, "ORDER_EXISTS");
            }
            if (executionWouldOverflow(c)) {
                return Result.reject(c.request, "OVERFLOW");
            }
            Order taker = new Order(c, sequence++);
            NavigableMap<Long, LinkedHashMap<String, Order>> opposite = c.side.equals("B") ? asks : bids;
            List<Trade> trades = new ArrayList<>();
            while (taker.remaining > 0 && !opposite.isEmpty()) {
                long price = opposite.firstKey();
                if (c.side.equals("B") ? price > c.price : price < c.price) {
                    break;
                }
                Order maker = opposite.firstEntry().getValue().values().iterator().next();
                // 与现有 MatchingService 的 CANCEL_TAKER 顺序一致：先前有效成交保留，剩余量撤销。
                if (maker.owner.equals(c.owner)) {
                    taker.cancelled = taker.remaining;
                    taker.remaining = 0;
                    break;
                }
                long quantity = Math.min(taker.remaining, maker.remaining);
                // 每笔成交量不大于 maker 原始量，maker 的 price*original 已验证，不会货币溢出。
                maker.remaining -= quantity; maker.filled += quantity;
                taker.remaining -= quantity; taker.filled += quantity;
                long notional = quantity * maker.price;
                maker.executed += notional; taker.executed += notional;
                trades.add(new Trade(maker.id, quantity, maker.price));
                if (maker.remaining == 0) {
                    remove(maker);
                }
            }
            orders.put(taker.id, taker);
            if (taker.remaining > 0) {
                book(taker).computeIfAbsent(taker.price, ignored -> new LinkedHashMap<>()).put(taker.id, taker);
            }
            return new Result("A", c.request, c.id, taker.remaining, taker.cancelled, trades, "");
        }

        /** 卖单可能以高于限价成交；累积成交额与剩余挂单预留之和也必须可表示。 */
        boolean executionWouldOverflow(Command c) {
            var opposite = c.side.equals("B") ? asks : bids;
            long left = c.quantity, total = 0;
            outer: for (var level : opposite.entrySet()) {
                if (c.side.equals("B") ? level.getKey() > c.price : level.getKey() < c.price) {
                    break;
                }
                for (Order maker : level.getValue().values()) {
                    if (maker.owner.equals(c.owner)) {
                        left = 0;
                        break outer;
                    }
                    long quantity = Math.min(left, maker.remaining), notional = quantity * maker.price;
                    if (total > Long.MAX_VALUE - notional) {
                        return true;
                    }
                    total += notional; left -= quantity;
                    if (left == 0) {
                        break outer;
                    }
                }
            }
            return total > Long.MAX_VALUE - left * c.price;
        }

        NavigableMap<Long, LinkedHashMap<String, Order>> book(Order o) { return o.side.equals("B") ? bids : asks; }
        void remove(Order o) {
            var book = book(o); var level = book.get(o.price);
            level.remove(o.id);
            if (level.isEmpty()) {
                book.remove(o.price);
            }
        }

        /** 验证路径只用于重放；计时路径在整批结束后调用，避免把 O(N) 检查算入引擎延迟。 */
        void verify() {
            HashSet<Order> seen = new HashSet<>();
            checkBook(bids, "B", seen); checkBook(asks, "S", seen);
            for (Order o : orders.values()) {
                require(o.remaining >= 0 && o.filled >= 0 && o.cancelled >= 0, "negative quantity");
                require(o.original - o.remaining - o.filled == o.cancelled, "quantity conservation");
                require(o.executed >= 0 && o.executed <= Long.MAX_VALUE - o.remaining * o.price, "execution notional overflow");
                require((o.remaining > 0) == seen.contains(o), "order/book index disagreement");
            }
            require(bids.isEmpty() || asks.isEmpty() || bids.firstKey() < asks.firstKey(), "crossed book");
        }
        void checkBook(NavigableMap<Long, LinkedHashMap<String, Order>> book, String side, HashSet<Order> seen) {
            for (var entry : book.entrySet()) {
                require(!entry.getValue().isEmpty(), "empty level"); long previous = -1;
                for (Order o : entry.getValue().values()) {
                    require(o.price == entry.getKey() && o.side.equals(side) && o.remaining > 0, "bad level");
                    require(o.sequence > previous && seen.add(o) && orders.get(o.id) == o, "FIFO/index"); previous = o.sequence;
                }
            }
        }
        String snapshot() {
            StringBuilder b = new StringBuilder();
            for (Order o : new TreeMap<>(orders).values()) {
                if (!b.isEmpty()) {
                    b.append(';');
                }
                b.append(o.id).append(':').append(o.owner).append(':').append(o.side).append(':').append(o.price)
                 .append(':').append(o.original).append(':').append(o.remaining).append(':').append(o.filled).append(':').append(o.cancelled).append(':').append(o.executed);
            }
            return b.toString();
        }
    }

    static void require(boolean ok, String reason) {
        if (!ok) {
            throw new IllegalStateException(reason);
        }
    }
    static String hash(String state) {
        long h = 0xcbf29ce484222325L;
        for (byte b : state.getBytes(StandardCharsets.US_ASCII)) {
            h = (h ^ (b & 255)) * 0x100000001b3L;
        }
        // 这是状态散列的有意模 2^64 运算，不参与任何金额计算。
        return Long.toUnsignedString(h);
    }
    static long percentile(long[] sorted, double p) { return sorted[Math.max(0, (int) Math.ceil(sorted.length * p) - 1)]; }
    static volatile long observable;

    public static void main(String[] args) throws Exception {
        if (args.length == 1 && (args[0].equals("--replay") || args[0].equals("--replay-batch"))) {
            Engine engine = new Engine();
            try (BufferedReader in = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8))) {
                for (String line; (line = in.readLine()) != null;) {
                    System.out.println(engine.apply(Command.parse(line)).encode());
                    if (args[0].equals("--replay")) {
                        engine.verify();
                    }
                }
            }
            engine.verify(); System.out.println("STATE|" + engine.snapshot()); return;
        }
        if (args.length != 3 || !args[0].equals("--bench")) {
            throw new IllegalArgumentException("--replay | --bench FILE WARMUP");
        }
        Path input = Path.of(args[1]);
        List<Command> commands = Files.readAllLines(input).stream().map(Command::parse).toList();
        int warmup = Integer.parseInt(args[2]);
        require(!commands.isEmpty() && warmup > 0, "empty workload or missing warmup");
        for (int pass = 0; pass < warmup; pass++) {
            Engine warm = new Engine();
            for (Command c : commands) {
                warm.apply(c);
            }
            warm.verify(); observable ^= warm.snapshot().length();
        }
        Engine engine = new Engine(); long[] samples = new long[commands.size()];
        var bean = ManagementFactory.getThreadMXBean();
        com.sun.management.ThreadMXBean allocation = bean instanceof com.sun.management.ThreadMXBean a && a.isThreadAllocatedMemorySupported() ? a : null;
        if (allocation != null && !allocation.isThreadAllocatedMemoryEnabled()) {
            allocation.setThreadAllocatedMemoryEnabled(true);
        }
        long thread = Thread.currentThread().threadId();
        long allocatedBefore = allocation == null ? -1 : allocation.getThreadAllocatedBytes(thread);
        long started = System.nanoTime();
        for (int i = 0; i < commands.size(); i++) {
            Command c = commands.get(i); long before = System.nanoTime();
            engine.apply(c); samples[i] = System.nanoTime() - before;
        }
        long elapsed = System.nanoTime() - started;
        long allocated = allocation == null ? -1 : allocation.getThreadAllocatedBytes(thread) - allocatedBefore;
        engine.verify(); String stateHash = hash(engine.snapshot()); observable ^= stateHash.length();
        // 原始样本在计时完成后落盘，保留每个独立进程的分布供复核。
        Path sampleFile = input.resolveSibling(input.getFileName() + ".java.samples." + ProcessHandle.current().pid() + ".csv");
        try (var writer = Files.newBufferedWriter(sampleFile)) {
            writer.write("latency_ns\n");
            for (long sample : samples) {
                writer.write(sample + "\n");
            }
        }
        Arrays.sort(samples);
        System.out.println("{\"samples\":" + samples.length + ",\"elapsedNs\":" + elapsed + ",\"throughputOpsSec\":" + (samples.length * 1e9 / elapsed)
            + ",\"p50Ns\":" + percentile(samples, .5) + ",\"p99Ns\":" + percentile(samples, .99) + ",\"p999Ns\":" + percentile(samples, .999)
            + ",\"maxNs\":" + samples[samples.length - 1] + ",\"allocatedBytesMeasuredThread\":" + (allocated < 0 ? "null" : allocated)
            + ",\"stateHash\":\"" + stateHash + "\",\"sampleFile\":\"" + sampleFile.getFileName() + "\"}");
    }
}
